adapters/src/main/kotlin/com/krzykrucz/elesson/currentlesson/preparedness/adapters/readmodel/PreparednessInSemesterReadModel.kt

```diff
@@ -0,0 +1,61 @@
+package com.krzykrucz.elesson.currentlesson.preparedness.adapters.readmodel
+
+import arrow.core.Option
+import arrow.core.getOrElse
+import arrow.effects.IO
+import com.krzykrucz.elesson.currentlesson.preparedness.adapters.readmodel.StudentInSemesterReadModel.READ_MODEL
+import com.krzykrucz.elesson.currentlesson.preparedness.domain.StudentMarkedUnprepared
+import com.krzykrucz.elesson.currentlesson.preparedness.readmodel.GetStudentSubjectUnpreparednessInASemester
+import com.krzykrucz.elesson.currentlesson.preparedness.readmodel.StudentInSemester
+import com.krzykrucz.elesson.currentlesson.preparedness.readmodel.StudentSubjectUnpreparednessInASemester
+import com.krzykrucz.elesson.currentlesson.preparedness.readmodel.WriteUnpreparednessInTheRegister
+import com.krzykrucz.elesson.currentlesson.shared.AsyncFactory
+import org.springframework.context.annotation.Bean
+import org.springframework.context.annotation.Configuration
+import org.springframework.context.event.EventListener
+import java.util.concurrent.ConcurrentHashMap
+
+class ReadModelProjector : WriteUnpreparednessInTheRegister {
+
+    @EventListener
+    override fun invoke(event: StudentMarkedUnprepared) =
+        StudentInSemester(
+            event.lessonId.className,
+            event.unpreparedStudent.firstName,
+            event.unpreparedStudent.secondName
+        )
+            .let {
+                Option.fromNullable(READ_MODEL[it])
+                    .getOrElse { StudentSubjectUnpreparednessInASemester.createEmpty(it) }
+            }
+            .let {
+                READ_MODEL.compute(it.student) { student, unprep ->
+                    Option.fromNullable(unprep)
+                        .getOrElse { StudentSubjectUnpreparednessInASemester.createEmpty(student) }
+                        .let { it.copy(count = it.count.inc()) }
+                } ?: it
```

Comment:
guys, I have, maybe wrong, impression but pipelining everything without providing meaningful, well-named steps of the pipeline looks ugly and unreadable. It would look much better it operations in `let` were extracted to smaller functions.

Reply:
The first `let` is ok imo, but I'd extract the second one for sure.